### PR TITLE
Add ability to modify a material's charge

### PR DESCRIPTION
**Pull request by psavery**
Note that we have to update the unit cell and structure factor
manually [since the material is not currently doing that](https://github.com/HEXRD/hexrd/blob/300f43ffe35cdbd0a7827b80d0b9209f196771a0/hexrd/material.py#L864-L866).

Maybe we should just add the charge as an argument to
[`set_atomdata()`](https://github.com/HEXRD/hexrd/blob/300f43ffe35cdbd0a7827b80d0b9209f196771a0/hexrd/material.py#L942)?

Fixes: #886

Depends on: hexrd/hexrd#301

### CODE REVIEW

File: hexrd/ui/material_structure_editor.py [hunk 1]
```diff
@@ -249,17 +250,25 @@ def update_material(self):
...
                 U_array.append(atom['U'])
 
         mat = self.material
         mat._set_atomdata(type_array, info_array, U_array)
+        mat.charge = charge_array
+
+        # Must force an update of the unit cell and structure factor
+        # for the change in the charge array.
+        mat._newUnitcell()
+        mat.update_structure_factor()
```
Comment (psavery) [inline hexrd/ui/material_structure_editor.py:271]: Here is where we are updating the unit cell and structure factor manually.

**Comment by joelvbernier** [comment]
Where is this one at @saransh13 and @psavery ?

**Comment by psavery** [comment]
> Where is this one at @saransh13 and @psavery ?

Ready for @saransh13 to review.

**Review by saransh13 — COMMENTED** [review]
The following error is generated when I try to update the atomic species. For this example, I switched Nickel to Cobalt and got this error.

I have also pushed an update to `_set_atomdata`[https://github.com/HEXRD/hexrd/blob/1c25c6e70695010c8f1e55b5819a3d716ac44259/hexrd/material.py#L967] which takes the charge as an input. The corresponding changes to the guidelines should be here[https://github.com/HEXRD/hexrdgui/blob/5149e8eb7c059ff9876a41fa1b18caa80601e8f0/hexrd/ui/material_structure_editor.py#L262]. I tried to directly push the update to your fork, but I don't think I have the right permissions.

Screen Shot 2021-07-23 at 12 28 11 PM

**Comment by psavery** [comment]
> I have also pushed an update to [`_set_atomdata`](https://github.com/HEXRD/hexrd/blob/1c25c6e70695010c8f1e55b5819a3d716ac44259/hexrd/material.py#L967) which takes the charge as an input. The corresponding changes to the `hexrdgui` should be [here](https://github.com/HEXRD/hexrdgui/blob/5149e8eb7c059ff9876a41fa1b18caa80601e8f0/hexrd/ui/material_structure_editor.py#L262). I tried to directly push the update to your fork, but I don't think I have the right permissions.

It looks like [`_set_atomdata`](https://github.com/HEXRD/hexrd/blob/bf0d30c57c7497f99b8d8402a5261f104fa17e29/hexrd/material.py#L967) isn't updated yet on hexrd master with those changes. Can you put up a PR for it?

**Comment by psavery** [comment]
> The following error is generated when I try to update the atomic species. For this example, I switched Nickel to Cobalt and got this error.

I just fixed this. Thank you for finding it.

**Review by saransh13 — APPROVED** [review]
LGTM. The new PR for updating charge with other artomdata is up, so this could be updated to use that before merging. But otherwise this works really well.

File: hexrd/ui/material_structure_editor.py (hunk 1 above)
Comment (psavery) [reply hexrd/ui/material_structure_editor.py]: This code is now fixed to pass the charge via the changes in hexrd/hexrd#301.

**Comment by joelvbernier** [comment]
Should we merge HEXRD/hexrd/pull/301 and make the required changes here?  I can tag hexrd in that case.

**Comment by psavery** [comment]
> Should we merge [HEXRD/hexrd/pull/301](https://github.com/HEXRD/hexrd/pull/301) and make the required changes here? I can tag hexrd in that case.

The required changes are actually already in place in this PR! We just need to merge hexrd/hexrd#301 first, and then we can merge this one.

**Pull request merged**